### PR TITLE
.NET (Core?) CatBoost interface

### PR DESCRIPTION
**Pull request by 17minutes**
Here I would like to re-initiate the discussion started in the #409 (Cc @kizill)

In this PR I'm working on the .NET interface to the `catboostlib.dll`. So far, I have the following ready:

- Low-level model API which basically replicates DLL API but uses managed types (see `CatBoostNet` project);
- Console app (see `ProofOfConcept` project) using CatBoost `.cbm` models trained on the following datasets:
  * Iris (multi-classification, float only features)
  * Boston housing (regression, float only features)
  * Mushroom (binary classification, _cat_ only features)

Obviously, this is WIP, so any relevant ideas are appreciated. However, I would also like to ask several questions that are of specific interest to me:

1. Can I somehow get indices of the categorical features from the model using DLL API — or, rather, should I provide my own method in the DLL for that? (So far, I'm providing such indices manually, which is clearly a wrong approach.)
2. Similarly, can I get feature names form the model with DLL API or should I provide my own method?
3. How can I debug unmanaged API calls from the managed code? While I did manage to get the current API working, I'm not sure how would I handle more complex problems without looking at the calls / memory / etc.
4. As a continuation of (3), are there any known issues around calling `GetErrorString`? When I run it from my `CatBoostModelEvaluator` class, both the program and the debug session terminate _without throwing any exceptions_.
5. I have the following idea for the demo: instead of making `.ipynb` showcasing library use in .NET language (be it C# or F#), train CatBoost model on some reasonable real-world dataset and show the F# micro-server running that model. What do you think about it, and if this is a reasonable idea, can you share your opinion on the possible dataset?

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=ru

### CODE REVIEW

File: ya.bat [hunk 1]
```diff
@@ -8,13 +8,13 @@ call :dbg Ya: %YA_BAT_REAL%
 call :find_python
 if ERRORLEVEL 1 exit /b 1
 call :dbg Python: %YA_BAT_PYTHON%
-call %YA_BAT_PYTHON% %YA_BAT_REAL% %*
+call %YA_BAT_PYTHON% "%YA_BAT_REAL%" %*
```
Comment (exprmntr) [inline ya.bat]: Please make separate pull request with this change

File: catboost/dotnet/.gitignore [hunk 2]
```diff
@@ -0,0 +1,349 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (kizill) [inline catboost/dotnet/.gitignore]: Есть ощущение, что это слишком широкие гитигноры. Но в любом случае, .gitignore должен лежать в корне репозитория. Давай уберем .gitignore файлы из ревью?

**Comment by 17minutes** [comment]
> catboost/dotnet/CatBoostNet/catboostmodel.dll - бинарные артефакты в репозиторий класть не хорошо, возможно правильнее собирать этот файл локально. Лучше в солюшне сделать проект с CustomBuildStep-ом, который вызывает ya make catboost/libs/model_interface -o <dll_target_path>
> Еще нужно удалить датасеты (как минимум boston housing), лучше для мелких примеров брать данные из https://github.com/catboost/catboost/tree/master/catboost/pytest/data, большие примеры лучше скачивать из сети, например при помощи https://github.com/catboost/catboost/blob/master/catboost/python-package/catboost/datasets.py , либо качать датасеты из сети.

@kizill Всё перечисленное ⬆️ реализовал в последних коммитах (начиная с `
38adf29`) — и прошу прощения за отсутствие реакции, у меня сейчас идёт зачётная неделя... реагировать активнее @ продолжать работу над PR планирую в конце мая — начале июня (после окончания зачётного периода).

**Comment by kizill** [comment]
Ты так и не поправил замечания про ya.bat от Антона.

File: ya.bat (hunk 1 above)
Comment (17minutes) [reply ya.bat]: #829

**Comment by kizill** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

File: catboost/dotnet/paket.dependencies [hunk 3]
```diff
@@ -0,0 +1 @@
+source https://www.nuget.org/api/v2
```
Comment (annaveronika) [inline catboost/dotnet/paket.dependencies]: Please move the tutorial to catboost/tutorials/apply_model/dotnet

**Comment by kizill** [comment]
@arcadia-devtools Ship it!

**Comment by kizill** [comment]
@arcadia-devtools Ship it!

**Comment by annaveronika** [comment]
@17minutes Please write text for documentation (see examples here https://catboost.ai/docs/concepts/applying__models.html)

You can write text here, we'll add it to the site.

**Comment by kizill** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@kizill, internal review request created: [825587](https://a.yandex-team.ru/review/825587)

**Comment by arcadia-devtools** [comment]
Check [825587](https://a.yandex-team.ru/review/825587) passed
